x/ibc/02-client/client/cli/query.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/cosmos/cosmos-sdk/version"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types/tendermint"
+	ics23 "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment"
+	"github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/merkle"
+)
+
+// GetQueryCmd returns the query commands for IBC clients
+func GetQueryCmd(queryRouter string, cdc *codec.Codec) *cobra.Command {
+	ics02ClientQueryCmd := &cobra.Command{
+		Use:                        "client",
+		Short:                      "IBC client query subcommands",
+		DisableFlagParsing:         true,
+		SuggestionsMinimumDistance: 2,
+	}
+
+	ics02ClientQueryCmd.AddCommand(client.GetCommands(
+		GetCmdQueryConsensusState(queryRouter, cdc),
+		GetCmdQueryPath(queryRouter, cdc),
+		GetCmdQueryHeader(cdc),
+		GetCmdQueryClientState(queryRouter, cdc),
+		GetCmdQueryRoot(queryRouter, cdc),
+	)...)
+	return ics02ClientQueryCmd
+}
+
+// GetCmdQueryClientState defines the command to query the state of a client with
+// a given id as defined in https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics#query
+func GetCmdQueryClientState(storeKey string, cdc *codec.Codec) *cobra.Command {
+	return &cobra.Command{
+		Use:   "state [client-id]",
+		Short: "Query a client state",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query stored client state
+
+Example:
+$ %s query ibc client state [client-id]
+		`, version.ClientName),
+		),
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+			id := args[0]
+
+			bz, err := cdc.MarshalJSON(types.NewQueryClientStateParams(id))
+			if err != nil {
+				return err
+			}
+
+			res, _, err := cliCtx.QueryWithData(types.ClientStatePath(id), bz)
+			if err != nil {
+				return err
+			}
+
+			var clientState types.ClientState
+			if err := cdc.UnmarshalJSON(res, &clientState); err != nil {
+				return err
+			}
+
+			return cliCtx.PrintOutput(clientState)
+		},
+	}
+}
+
+// GetCmdQueryRoot defines the command to query
+func GetCmdQueryRoot(storeKey string, cdc *codec.Codec) *cobra.Command {
+	return &cobra.Command{
+		Use:   "root [client-id] [height]",
+		Short: "Query stored root",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Query a stored commitment root at a specific height for a particular client
+
+Example:
+$ %s query ibc client root [client-id] [height]
+`, version.ClientName),
+		),
+		Args: cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+			id := args[0]
+			height, err := strconv.ParseUint(args[1], 10, 64)
+			if err != nil {
+				return err
+			}
+
+			bz, err := cdc.MarshalJSON(types.NewQueryCommitmentRootParams(id, height))
+			if err != nil {
+				return err
+			}
+
+			res, _, err := cliCtx.QueryWithData(types.RootPath(id, height), bz)
+			if err != nil {
+				return err
```

Comment:
Return proper error if client-id isn't found. Also do we want to do any input validation here?

Reply:
> Return proper error if `client-id` at `height` isn't found.

this is handled by `Args: cobra.ExactArgs(2),`